app/client/rest/tracking.ts

```diff
@@ -93,40 +99,118 @@ export default class ClientTracking {
         group.totalCompressedSize += metrics?.compressedSize ?? 0;
     }
 
-    incrementRequestCount(groupLabel: string) {
+    getAverageLatency(groupLabel: RequestGroupLabel): number {
+        const groupData = this.requestGroups.get(groupLabel);
+        if (!groupData) {
+            return 100;
+        }
+
+        const urlData = Object.entries(groupData.urls);
+        const sumLatency = urlData.reduce((result, url) => (result + (url[1].metrics?.latency ?? 0)), 0);
+        return sumLatency / urlData.length;
+    }
+
+    categorizeRequests(groupLabel: RequestGroupLabel): CategorizedRequests {
+        const parallel: ClientResponseMetrics[] = [];
+        const sequential: ClientResponseMetrics[] = [];
+        const groupData = this.requestGroups.get(groupLabel);
+        if (!groupData) {
+            return {parallel, sequential};
+        }
+
+        // Sort requests by start time
+        const requestsMetrics = Object.entries(groupData.urls).map((e) => e[1].metrics).filter((m) => m != null);
+        requestsMetrics.sort((a, b) => a.startTime - b.startTime);
+
+        let lastEndTime = 0;
+
+        for (const metrics of requestsMetrics) {
+            if (metrics.startTime < lastEndTime) {
+                // Overlapping request -> Parallel
+                parallel.push(metrics);
+            } else {
+                // Non-overlapping request -> Sequential
+                sequential.push(metrics);
+            }
+
+            // Update the last end time
+            lastEndTime = Math.max(lastEndTime, metrics.endTime);
+        }
+
+        return {parallel, sequential};
+    }
+
+    calculateAverageSpeedWithCategories = (
+        categorizedRequests: CategorizedRequests,
+        elapsedTimeInSeconds: number, // Observed total elapsed time in seconds
+    ): { averageSpeedMbps: number; effectiveLatency: number } => {
+        const {parallel, sequential} = categorizedRequests;
+
+        // Step 1: Calculate total data size in bits
+        const totalDataBits = [...parallel, ...sequential].reduce((sum, req) => sum + (req.compressedSize * 8), 0);
+
+        // Step 2: Calculate effective latency
+        const parallelLatencies = parallel.map((req) => req.latency);
+        const sequentialLatencies = sequential.map((req) => req.latency);
+
+        const effectiveParallelLatency = Math.max(...parallelLatencies, 0);
+        const effectiveSequentialLatency = sequentialLatencies.reduce((sum, latency) => sum + latency, 0);
+
+        const effectiveLatency = effectiveParallelLatency + effectiveSequentialLatency;
```

Comment:
Do these need to be summed or should we get the avg especially for sequential latency?

Reply:
as far as I'm aware, for parallel requests the latency is the maximum latency of all requests and then sequential requests do add up to get the effective latency as if you execute two sequential requests and one takes 3s and the other one 5s then you waited a total of 8s.

that is the reason why I added the avg latency overall.